src/lxml/html/diff.py

```diff
@@ -234,22 +234,38 @@ def expand_tokens(tokens, equal=False):
 def merge_insert(ins_chunks, doc):
     """ doc is the already-handled document (as a list of text chunks);
     here we add <ins>ins_chunks</ins> to the end of that.  """
-    # Though we don't throw away unbalanced_start or unbalanced_end
-    # (we assume there is accompanying markup later or earlier in the
-    # document), we only put <ins> around the balanced portion.
+
+    ins_chunks = list(ins_chunks)
     unbalanced_start, balanced, unbalanced_end = split_unbalanced(ins_chunks)
-    doc.extend(unbalanced_start)
+    unbalanced_tags = unbalanced_start + unbalanced_end
+
     if doc and not doc[-1].endswith(' '):
         # Fix up the case where the word before the insert didn't end with 
         # a space
         doc[-1] += ' '
     doc.append('<ins>')
-    if balanced and balanced[-1].endswith(' '):
-        # We move space outside of </ins>
-        balanced[-1] = balanced[-1][:-1]
-    doc.extend(balanced)
-    doc.append('</ins> ')
-    doc.extend(unbalanced_end)
+
+    for chunk in ins_chunks:
+        if chunk in balanced:
+            doc.append(chunk)
```

Comment:
This seems risky. I'm not sure that we can assume chunks to be unique enough to just do an "in" test here.

Maybe it could help to put the three parts back together, but give each chunk a marker like `(chunk, "balanced")` and `(chunk, "unbalanced")`, and then do something based on the marker. Or look at the index boundaries where we switch between the unbalanced and balanced parts.

Reply:
Hmm, agreed that we can't assume the chunks to be unique enough to just do an 'in' test here. I'll think about your suggestion and give this another go. Thanks for your review.

Reply:
We ran into an issue rising out of precisely the problem you caught ! The visible effect being that the `<ins>` tags were not being added at places where they ought to be.

I've updated the approach by changing `split_unbalanced` to *optionally* return a list of tuples containing the `(<index-in-the-input=chunks-list>, <chunk>)`. This then serves to uniquely identify the chunk across the balanced/unbalanced lists ...which then makes the `in` operator reliable.

I've defaulted to retain the existing behavior of `split_unbalanced` since it is also called from `cleanup_delete` and I didn't want to touch that function.

Hopefully, you find the changes suitable.